component/loki/source/azureeventhubs/azureeventhubs.go

```diff
@@ -0,0 +1,210 @@
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the loki.source.azure_event_hubs component.
+type Arguments struct {
+	FullyQualifiedNamespace string   `river:"fully_qualified_namespace,attr"`
+	EventHubs               []string `river:"event_hubs,attr"`
+
+	Authentication AzureEventHubsAuthentication `river:"authentication,block"`
+
+	GroupID                string             `river:"group_id,attr,optional"`
+	UseIncomingTimestamp   bool               `river:"use_incoming_timestamp,attr,optional"`
+	DisallowCustomMessages bool               `river:"disallow_custom_messages,attr,optional"`
+	RelabelRules           flow_relabel.Rules `river:"relabel_rules,attr,optional"`
+	Labels                 map[string]string  `river:"labels,attr,optional"`
+	Assignor               string             `river:"assignor,attr,optional"`
+
+	ForwardTo []loki.LogsReceiver `river:"forward_to,attr"`
+}
+
+// AzureEventHubsAuthentication describe the configuration for authentication with Azure Event Hub
+type AzureEventHubsAuthentication struct {
+	Mechanism        string   `river:"mechanism,attr"`
+	Scopes           []string `river:"scopes,attr,optional"`
+	ConnectionString string   `river:"connection_string,attr,optional"`
+}
+
+func getDefault() Arguments {
+	return Arguments{
+		GroupID:  "loki.source.azure_event_hubs",
+		Labels:   map[string]string{"job": "loki.source.azure_event_hubs"},
+		Assignor: "range",
+	}
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (a *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*a = getDefault()
+	type arguments Arguments
+	return f((*arguments)(a))
+}
+
+// New creates a new loki.source.azure_event_hubs component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		mut:     sync.RWMutex{},
+		opts:    o,
+		handler: make(loki.LogsReceiver),
+		fanout:  args.ForwardTo,
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Component implements the loki.source.azure_event_hubs component.
+type Component struct {
+	opts    component.Options
+	mut     sync.RWMutex
+	fanout  []loki.LogsReceiver
+	handler loki.LogsReceiver
+	target  *kt.TargetSyncer
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.azure_event_hubs component shutting down, stopping the targets")
+		c.mut.RLock()
+		err := c.target.Stop()
+		if err != nil {
+			level.Error(c.opts.Logger).Log("msg", "error while stopping azure_event_hubs target", "err", err)
+		}
+		c.mut.RUnlock()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			c.mut.RLock()
```

Comment:
for @tpaschalis: we might be susceptible to something like #3391 here, both for this and the loki.source.kafka component.

Will check it out to see if there's a need for a fix after this is merged.